mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -397,49 +436,132 @@ object GeneralizedLinearRegression extends DefaultParamsReadable[GeneralizedLine
 
     /** Trim the fitted value so that it will be in valid range. */
     def project(mu: Double): Double = mu
+
   }
 
   private[regression] object Family {
 
     /**
-     * Gets the [[Family]] object from its name.
+     * Gets the [[Family]] object based on family and variancePower.
+     * 1) retrieve object based on family name
+     * 2) if family name is tweedie, retrieve object based on variancePower
      *
-     * @param name family name: "gaussian", "binomial", "poisson" or "gamma".
+     * @param model a GenerealizedLinearRegressionBase object
      */
-    def fromName(name: String): Family = {
-      name match {
-        case Gaussian.name => Gaussian
-        case Binomial.name => Binomial
-        case Poisson.name => Poisson
-        case Gamma.name => Gamma
+    def fromModel(model: GeneralizedLinearRegressionBase): Family = {
+      model.getFamily match {
+        case "gaussian" => Gaussian
+        case "binomial" => Binomial
+        case "poisson" => Poisson
+        case "gamma" => Gamma
+        case "tweedie" =>
+          model.getVariancePower match {
+            case 0.0 => Gaussian
+            case 1.0 => Poisson
+            case 2.0 => Gamma
+            case default => new TweedieFamily(default)
+          }
       }
     }
   }
 
   /**
-   * Gaussian exponential family distribution.
-   * The default link for the Gaussian family is the identity link.
-   */
-  private[regression] object Gaussian extends Family("gaussian") {
+    * Tweedie exponential family distribution.
+    * This includes the special cases of Gaussian, Poisson and Gamma.
+    */
+  private[regression] class TweedieFamily(private val variancePower: Double)
+    extends Family{
+
+    val name: String = variancePower match {
```

Comment:
Why remove the name and switch like this? you can instead adjust this so that subclasses override a `name` method.